Async DNS over TCP

raboof opened the pull request:
Refs #25460

* [x] Basic functionality
* [x] Retry/reconnect logic on failures
* [x] Buffer/reassemble fragmented replies

The current implementation includes an integration test that relies on being online and receiving a truncated response when resolving the `AAAA` records for `many.bzzt.net`. We're considering creating a Docker image that acts as a DNS server that shows the behavior we want to test against.

raboof committed:
Async DNS over TCP (#25460)

chbatey commented on akka-actor-tests/src/test/scala/akka/io/dns/OnlineAsyncDnsResolverIntegrationSpec.scala:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    "resolve queries that are too big for UDP" in {
+      val name = "many.bzzt.net"
```
is this still up? Not working for me

raboof replied:
Should! (serving AAAA records)

raboof committed:
Reconnect after getting disconnected

But not right away, as apparently this DNS server doesn't like open
connections.

raboof committed:
Restart with backoff on Connect or Write failures

raboof committed:
Reassemble partial responses

raboof commented:
(cancelled 2 builds to make some room on Jenkins, we're mostly interested in the newest commit after all)

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala:
```diff
@@ -40,7 +42,15 @@ import scala.util.Try
+    minBackoff = 10.millis,
+    maxBackoff = 20.seconds,
+    randomFactor = 0.1
+  ), "tcpDnsClientSupervisor")
```
I'm guessing this only happens for the SRV use case, not normal DNS, so should we perhaps start it lazily instead of always?

raboof replied:
In theory it can also happen for large normal DNS responses, but I agree that is a corner case and it'd make sense to start this lazily

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala:
```diff
@@ -105,18 +116,25 @@ import scala.util.Try
       // TODO remove me when #25460 is implemented
       if (msg.flags.isTruncated) {
-        log.warning("DNS response truncated and fallback to TCP is not yet implemented. See #25460")
+        log.debug("DNS response truncated, falling back to TCP")
```
remove comment above

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala:
```diff
@@ -105,18 +116,25 @@ import scala.util.Try
+        }
+      } else {
+        val (recs, additionalRecs) = if (msg.flags.responseCode == ResponseCode.SUCCESS) (msg.answerRecs, msg.additionalRecs) else (Nil, Nil)
+        self ! Answer(msg.id, recs, additionalRecs)
```
Why do we need to reschedule it a message, couldn't we just deal with the result right away?

raboof replied:
in the truncated case, it's the `TcpDnsClient` that sends the `Answer` to the `DnsClient`

chbatey replied:
Could it not get the replyTo out of the inflightRequests then send it back right away?

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala:
```diff
@@ -0,0 +1,98 @@
+  val connecting: Receive = {
+    case CommandFailed(_: Connect) ⇒
+      log.warning("Failed to connect to [{}]", ns)
+      throw new AkkaException("Connecting failed")
```
Isn't just throwing going to lead to it getting logged? Put the `ns` in the exception instead?

raboof replied:
Probably, but that might not be bad? Weird if connecting to the DNS server fails.

johanandren replied:
Mostly thinking that if we get two errors in the log that is confusing but maybe it doesn't matter.

raboof replied:
Ah, good one, will remove the `warning` :+1:

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala:
```diff
@@ -0,0 +1,98 @@
+
+  private def parseResponse(data: ByteString) = {
+    val msg = Message.parse(data)
+    log.debug(s"Decoded: $msg")
```
`debug("Decoded: {}", msg)`

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala:
```diff
@@ -0,0 +1,98 @@
+    case msg: Message ⇒
+      val bytes = msg.write()
+      connection ! Tcp.Write(encodeLength(bytes.length))
+      connection ! Tcp.Write(bytes)
```
Do them with one write message?

johanandren commented on akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala:
```diff
@@ -0,0 +1,98 @@
+        if (data.drop(prefixSize).length < expectedPayloadLength)
+          context.become(ready(connection, data))
+        else {
+          val payload = data.drop(prefixSize).take(expectedPayloadLength)
```
Isn't there something with drop and take on ByteStrings being overly expensive?

raboof replied:
Since we're doing I/O anyway and we need to seek in the message while parsing references to domain names, I think the `drop` is worth it. The `take` we can skip, though.

johanandren reviewed (commented):
Just a partial review...

raboof committed:
Lazily create the TCP DNS client, add tests for DnsClient

raboof committed:
Test with dockerized bind

raboof committed:
Avoid chown on the files on the mounted volume

raboof committed:
Skip AsyncDnsResolverIntegrationSpec when docker missing

raboof committed:
Scalariform

raboof commented:
PLS BUILD

johanandren reviewed (approved):
LGTM

chbatey commented on akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala:
```diff
@@ -0,0 +1,92 @@
+
+  val idle: Receive = {
+    case _: Message ⇒
+      stash()
```
I guess the stash will never get too big if the dns server doesn't support TCP as we'll restart back a backoff

chbatey reviewed (approved):
Nice, looks good

patriknw commented on akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala:
```diff
@@ -0,0 +1,92 @@
+      val bytes = msg.write()
+      connection ! Tcp.Write(encodeLength(bytes.length) ++ bytes)
+    case CommandFailed(_: Write) ⇒
+      throw new AkkaException("Write failed")
```
does it include any reason for the failure that we want to have logged?

patriknw reviewed (approved):
LGTM

raboof committed:
Include cause in exception if possible
